File: html/js/pbclass/PbProject.js
// PbProject.js — port of Java PbProject.java (browser-compatible)
// Receives pre-extracted files as Map<string, Uint8Array> instead of reading disk.
import { PbFile } from './PbFile.js';
import { PbEntry } from './PbEntry.js';
import { PbEnum } from './PbEnum.js';

async function _decompressGzip(data) {
    const ds = new DecompressionStream('gzip');
    const writer = ds.writable.getWriter();
    const reader = ds.readable.getReader();
    writer.write(data);
    writer.close();
    const chunks = [];
    while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        chunks.push(value);
    }
    const total = chunks.reduce((n, c) => n + c.length, 0);
    const out = new Uint8Array(total);
    let off = 0;
    for (const c of chunks) { out.set(c, off); off += c.length; }
    return out;
}

const _latin1Dec = new TextDecoder('latin1');
const _utf16Dec  = new TextDecoder('utf-16le');

export class PbProject {
    constructor() {
        /** @type {PbFile[]} */
        this.files       = [];
        /** @type {Map<string, import('./PbObject.js').PbObject>} */
        this.objects     = new Map();
        /** @type {Map<number, import('./PbType.js').PbType>} */
        this.systemTypes = new Map();
        /** @type {Map<number, PbEnum>} */
        this.enums       = new Map();
        /** @type {import('./PbEntry.js').PbEntry|null} */
        this.systemEntry = null;
        /** @type {boolean} */
        this.isUnicode   = false;
        /** @type {boolean} */
        this.isPb5       = false;
        /** @type {number} */
        this.version     = 0;
        /** @type {boolean} */
        this.isDebug     = false;

        // Internal set to deduplicate library paths already processed
        this._loadedPaths = new Set();
        // Queue of {fileName, data} pairs added while iterating
        this._fileQueue   = [];
    }

    /**
     * Main async factory.
     * @param {string} entryFileName   — lowercased filename of the entry file
     * @param {Map<string, Uint8Array>} allFiles — Map<lowercased-name, Uint8Array>
     * @returns {Promise<PbProject>}
     */
    static async create(entryFileName, allFiles) {
        const project = new PbProject();
        project._allFiles = allFiles;

        // Find the entry file in allFiles
        const entryKey = entryFileName.toLowerCase();
        const entryData = allFiles.get(entryKey);
        if (!entryData) {
            throw new Error(`Entry file not found: ${entryFileName}`);
        }

        // Parse the entry PbFile (this may queue more files via onNewLibrary)
        const entryFile = new PbFile(project, entryFileName, entryData);
        project.files.push(entryFile);
        project._loadedPaths.add(entryKey);

        // Drain the queue — files added by onNewLibrary / onSystemLibrary
        let i = 0;
        while (i < project.files.length) {
            for (const entry of project.files[i].entries) {
                entry.parseObject();
            }
            i++;

            // Process any files added to the queue during parsing
            for (const { fileName, data } of project._fileQueue) {
                project.files.push(new PbFile(project, fileName, data));
            }
            project._fileQueue = [];
        }

        // Load system library (built-in function definitions) before parseInherit
        if (project.version > 0 && project.systemEntry === null) {
            await project._loadSystemLibrary();
        }

        if (project.systemEntry !== null) project.systemEntry.parseInherit();

        for (const pbFile of project.files) {
            for (const entry of pbFile.entries) {
                entry.parseInherit();
            }
        }

        return project;
    }

    // ── Callbacks from PbEntry / PbFile ──────────────────────────────────────

    /**
     * Called when a library path is encountered.
     * @param {string} libpath — bare filename (not full path)
     * @param {boolean} isFullPath — ignored in browser (we use allFiles map)
     */
    onNewLibrary(libpath, isFullPath) {
        const key = libpath.toLowerCase();
        if (this._loadedPaths.has(key)) return;
        this._loadedPaths.add(key);

        const data = this._allFiles ? this._allFiles.get(key) : undefined;
        if (data) {
            // queue for after current iteration
            this._fileQueue.push({ fileName: libpath, data });
        }
    }

    /**
     * Called when a version code is seen in a pbd entry header.
     * @param {number} version
     */
    onSystemLibrary(version) {
        if (this.version === 0) {
            this.version = version;
        } else if (this.version !== version) {
            console.warn('two version libraries in one project??');
        }
    }

    /** @param {import('./PbEntry.js').PbEntry} pbEntry */
    onSystemEntry(pbEntry) {
        this.systemEntry = pbEntry;
    }

    /** @param {import('./PbType.js').PbType} pbType */
    onNewSystemType(pbType) {
        this.systemTypes.set(pbType.index, pbType);
    }

    /**
     * @param {import('./PbObject.js').PbObject} pbObject
     * @param {string|null} [name]
     */
    onNewObject(pbObject, name = null) {
        this.objects.set(name !== null ? name : pbObject.type.name, pbObject);
    }

    /**
     * @param {import('./PbType.js').PbType} type
     * @param {number} index
     * @param {string} itemName
     */
    onNewEnumItem(type, index, itemName) {
        if (!this.enums.has(type.index)) {
            const e = new PbEnum();
            e.index = type.index;
            e.name  = type.name;
            this.enums.set(type.index, e);
        }
        this.enums.get(type.index).items.set(index, itemName + '!');
    }

    // ── System library loading ────────────────────────────────────────────────

    async _loadSystemLibrary() {
        const hex = this.version.toString(16).padStart(4, '0');
        try {
            const resp = await fetch(`data/resoures/${hex}.bin`);
            if (!resp.ok) return;
            const compressed = new Uint8Array(await resp.arrayBuffer());
            const decompressed = await _decompressGzip(compressed);
            const sysFile = PbFile.createSystem(this, this.version);
            const entry = new PbEntry(sysFile, '_typedef.grp', decompressed);
            sysFile.entries.push(entry);
            this.files.push(sysFile);
        } catch (e) {
            console.warn('Could not load system library:', e);
        }
    }

    // ── String helpers ────────────────────────────────────────────────────────

    /**
     * @param {Uint8Array} buffer
     * @param {number} [offset]
     * @param {number} [size]
     * @returns {string}
     */
    getString(buffer, offset = 0, size = buffer.length - offset) {
        const slice = buffer.slice(offset, offset + size);
        return this.isUnicode ? _utf16Dec.decode(slice) : _latin1Dec.decode(slice);
    }
}

File: java/src/main/java/com/pbdviewer/utils/pbclass/PbProject.java
package com.pbdviewer.utils.pbclass;

import java.io.File;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PbProject {

    private final String filePath;
    private final String dir;
    public boolean isDebug;

    private final List<PbFile> files = new ArrayList<>();
    private final Map<String, PbObject> objects = new HashMap<>();
    private final Map<Integer, PbType> systemTypes = new HashMap<>();
    private final Map<Integer, PbEnum> enums = new HashMap<>();
    private PbEntry systemEntry;
    private boolean isUnicode;
    private boolean isPb5;
    private int version;

    public PbProject(String filePath) {
        this.filePath = filePath;
        this.dir = new File(filePath).getParent();
        PbFile item = new PbFile(this, filePath);
        files.add(0, item);
        for (int i = 0; i < files.size(); i++) {
            for (PbEntry entry : files.get(i).getEntries()) {
                entry.parseObject();
            }
        }
        if (systemEntry != null) systemEntry.parseInherit();
        for (int i = 0; i < files.size(); i++) {
            for (PbEntry entry : files.get(i).getEntries()) {
                entry.parseInherit();
            }
        }
    }

    public String getString(byte[] buffer, int offset, int size) {
        if (!isUnicode) return new String(buffer, offset, size, Charset.defaultCharset());
        return new String(buffer, offset, size, StandardCharsets.UTF_16LE);
    }

    public String getString(byte[] buffer) {
        if (!isUnicode) return new String(buffer, Charset.defaultCharset());
        return new String(buffer, StandardCharsets.UTF_16LE);
    }

    public void onNewLibrary(String libpath, boolean isFullPath) {
        if (!isFullPath) libpath = Paths.get(dir, libpath).toString();
        if (!filePath.equalsIgnoreCase(libpath) && new File(libpath).exists()) {
            files.add(new PbFile(this, libpath));
        }
    }

    public PbFile onSystemLibrary(int version) {
        if (this.version == 0) {
            this.version = version;
            return new PbFile(this, this.version);
        }
        if (this.version != version) throw new RuntimeException("two version library in one project??");
        return null;
    }

    public void onSystemEntry(PbEntry pbEntry) {
        this.systemEntry = pbEntry;
    }

    public void onNewSystemType(PbType pbType) {
        systemTypes.put(pbType.getIndex(), pbType);
    }

    public void onNewObject(PbObject pbObject, String name) {
        objects.put(name != null ? name : pbObject.getType().getName(), pbObject);
    }

    public void onNewObject(PbObject pbObject) {
        onNewObject(pbObject, null);
    }

    public void onNewEnumItem(PbType type, int index, String itemName) {
        if (!enums.containsKey(type.getIndex())) {
            PbEnum pbEnum = new PbEnum();
            pbEnum.setIndex(type.getIndex());
            pbEnum.setName(type.getName());
            enums.put(type.getIndex(), pbEnum);
        }
        enums.get(type.getIndex()).getItems().put(index, itemName + "!");
    }

    // -------------------------------------------------------------------------
    // Getters and setters
    // -------------------------------------------------------------------------

    public List<PbFile> getFiles() {
        return files;
    }

    public Map<String, PbObject> getObjects() {
        return objects;
    }

    public Map<Integer, PbType> getSystemTypes() {
        return systemTypes;
    }

    public Map<Integer, PbEnum> getEnums() {
        return enums;
    }

    public PbEntry getSystemEntry() {
        return systemEntry;
    }

    public void setSystemEntry(PbEntry systemEntry) {
        this.systemEntry = systemEntry;
    }

    public boolean isUnicode() {
        return isUnicode;
    }

    public void setUnicode(boolean unicode) {
        this.isUnicode = unicode;
    }

    public boolean isPb5() {
        return isPb5;
    }

    public void setPb5(boolean pb5) {
        this.isPb5 = pb5;
    }

    public int getVersion() {
        return version;
    }

    public void setVersion(int version) {
        this.version = version;
    }
}
